src/pymor/algorithms/eigs.py

```diff
@@ -0,0 +1,292 @@
+
+        H, Qs = QR_iteration(H, shifts)
+
+        V = V.lincomb(Qs.T)
+        f = V[k] * H[k, k - 1] + f * Qs[l - 1, k - 1]
+        V = V[:k]
+        H = H[:k, :k]
+
+    return ews[:k0], V.lincomb(evs[:, :k0].T)
+
+
+def arnoldi(A, E, l, b):
+    """Compute an Arnoldi factorization.
+
+    Computes matrices :math:`V_l` and :math:`H_l` and a vector :math:`f_l` such that
+
+    .. math::
+        A V_l = V_l H_l + f_l e_l^T.
+
+    Additionally it holds that :math:`V_l^T V_l` is the identity matrix and :math:`H_l`
+    is an upper Hessenberg matrix. If `E` is not `None` it holds
+
+    .. math::
+        E^{-1} A V_l = V_l H_l + f_l e_l^T.
+
+    Parameters
+    ----------
+    A
+        The |Operator| A.
+    E
+        The |Operator| E.
+    l
+        The length of the Arnoldi factorization.
+    b
+        A |VectorArray| which is used as the initial vector for the iteration.
+
+    Returns
+    -------
+    V
+        A |VectorArray| whose columns span an orthogonal basis for R^l.
+    H
+        A |NumPy array| which is an upper Hessenberg matrix.
+    f
+        A |VectorArray| which represents the residual vector of the Arnoldi factorzation.
+    """
+
+    v = b * (1 / b.l2_norm()[0])
+
+    H = np.zeros((l, l))
+    V = A.source.empty(reserve=l)
+
+    V.append(v)
+
+    for i in range(l):
+        v = E.apply_inverse(A.apply(v))
+        V.append(v)
+
+        _, R = gram_schmidt(V, return_R=True, atol=0, rtol=0, offset=len(V) - 1, copy=False)
+        H[:i + 2, i] = R[:l, i + 1]
+        v = V[-1]
+
+    return V[:l], H, v * R[l, l]
+
+
+def extend_arnoldi(A, E, V, H, f, p):
+    """Extend an existing Arnoldi factorization.
+
+    Assuming that the inputs `V`, `H` and `f` define an Arnoldi factorization of length
+    :math:`l` (see :func:`arnoldi`), computes matrices :math:`V_{l+p}` and :math:`H_{l+p}`
+    and a vector :math:`f_{l+p}` which extend the factorization to a length `l+p` Arnoldi
+    factorization.
+
+    Parameters
+    ----------
+    A
+        The |Operator| A.
+    E
+        The |Operator| E.
+    V
+        The |VectorArray| V from the length :math:`l` Arnoldi factorization.
+    H
+        The |NumPy array| H from the length :math:`l` Arnoldi factorization.
+    f
+        The |VectorArray| f from the length :math:`l` Arnoldi factorization.
+    p
+        The number of addditional Arnoldi steps which are to be performed.
+
+    Returns
+    -------
+    V
+        A |VectorArray| whose columns span an orthogonal basis for R^(l+p).
+    H
+        A |NumPy array| which is an upper Hessenberg matrix.
+    f
+        A |VectorArray| which represents the residual vector of the Arnoldi factorzation.
+    """
+
+    k = len(V)
+
+    res = f.l2_norm()[0]
+    H = np.pad(H, ((0, p), (0, p)), 'constant')
+    H[k, k - 1] = res
+    v = f * (1 / res)
+    # since i cannot append to the VectorArrayView V I copy it before appending...
+    # is there a better way to do this?
+    V = V.copy()
+    V.append(v)
```

Comment:
Copying is ok here, I think. Generally, it's good not to mutate the input arguments without documenting this behavior.